packages/arb-avm-cpp/data_storage/src/timedsideloadcache.cpp

```diff
@@ -14,68 +14,70 @@
  * limitations under the License.
  */
 
-#include <data_storage/sideloadcache.hpp>
-
-size_t SideloadCache::size() {
-    std::shared_lock lock(mutex);
+#include <data_storage/timedsideloadcache.hpp>
 
+size_t TimedSideloadCache::size() {
     return cache.size();
 }
 
-void SideloadCache::add(std::unique_ptr<Machine> machine) {
-    std::unique_lock lock(mutex);
-
-    auto block_number = machine->machine_state.output.l2_block_number;
+void TimedSideloadCache::add(std::unique_ptr<Machine> machine) {
+    auto gas_used = machine->machine_state.output.arb_gas_used;
     auto timestamp = machine->machine_state.output.last_inbox_timestamp;
 
-    reorgNoLock(block_number);
-    deleteExpiredNoLock();
-
     if (timestamp <= expiredTimestamp()) {
         // Don't save expired machine to cache
         return;
     }
 
     // Add new entry
-    cache[block_number].timestamp = timestamp;
-    cache[block_number].machine = std::move(machine);
-}
+    cache[gas_used].timestamp = timestamp;
+    cache[gas_used].machine = std::move(machine);
 
-std::unique_ptr<Machine> SideloadCache::get(uint256_t block_number) {
-    std::shared_lock lock(mutex);
+    deleteExpired();
+}
 
-    auto it = cache.find(block_number);
-    if (it == cache.end()) {
-        return nullptr;
+std::optional<TimedSideloadCache::map_type::iterator>
+TimedSideloadCache::atOrBeforeGas(uint256_t gas_used) {
+    auto it = cache.upper_bound(gas_used);
+    if (it == cache.begin()) {
+        return std::nullopt;
     }
 
-    return std::make_unique<Machine>(*it->second.machine);
-}
-
-void SideloadCache::reorg(uint256_t next_block_number) {
-    std::unique_lock lock(mutex);
+    // Upper_bound returns the element after the one desired
+    it--;
 
-    reorgNoLock(next_block_number);
+    return it;
 }
 
-void SideloadCache::reorgNoLock(uint256_t next_block_number) {
-    if (next_block_number <= cache.begin()->first) {
+void TimedSideloadCache::reorg(uint256_t next_gas_used) {
+    if (next_gas_used <= cache.begin()->first) {
         // Remove everything
         cache.clear();
+
+        return;
     }
 
-    auto it = cache.lower_bound(next_block_number);
+    auto it = cache.lower_bound(next_gas_used);
     cache.erase(it, cache.end());
 }
 
-void SideloadCache::deleteExpiredNoLock() {
+void TimedSideloadCache::deleteExpired() {
     auto expired = expiredTimestamp();
+
     for (auto it = cache.cbegin();
          it != cache.cend() && it->second.timestamp <= expired;) {
         it = cache.erase(it);
     }
 }
 
-uint256_t SideloadCache::expiredTimestamp() const {
-    return std::time(nullptr) - expiration_seconds;
+uint256_t TimedSideloadCache::expiredTimestamp() {
+    if (cache.empty()) {
+        return 0;
+    }
+
+    return cache.crbegin()->second.timestamp - expiration_seconds;
```

Comment:
I'm not sure what exactly this logic is checking, but I don't think it's correct or at least correctly used within `timedsideloadcache.cpp`. Maybe some callsites there need `currentTimeExpired` instead.